File: docs/src/engine.js
const SORT_KEY = Object.freeze({
  TTK: 'ttk',
  DPS: 'dps',
  DAMAGE: 'damage',
  DAMAGE_END: 'damageend',
  FIRE_RATE: 'firerate',
  MAGAZINE: 'magazine',
});

const PRIORITY = Object.freeze({ HIGHEST: 'highest', LOWEST: 'lowest', AUTO: 'auto' });

export function normalizeData(raw) {
  if (raw && Number(raw.version) === 1 && Array.isArray(raw.cores)) {
    return normalizeWebData(raw);
  }

  return normalizeLegacyData(raw ?? {});
}

function normalizeWebData(raw) {
  return {
    cores: (raw.cores || []).map((core) => ({
      name: String(core.name || ''),
      category: String(core.category || ''),
      damage: Number(core.damage || 0),
      damageEnd: Number(core.damage_end || 0),
      fireRate: Number(core.fire_rate || 0),
    })),
    magazines: (raw.magazines || []).map((magazine) => ({
      name: String(magazine.name || ''),
      category: String(magazine.category || ''),
      magazineSize: Number(magazine.magazine_size || 0),
      damageMod: Number(magazine.damage_mod || 0),
      fireRateMod: Number(magazine.fire_rate_mod || 0),
    })),
    barrels: normalizeWebParts(raw.barrels || []),
    stocks: normalizeWebParts(raw.stocks || []),
    grips: normalizeWebParts(raw.grips || []),
    penalties: Array.isArray(raw.penalties) ? raw.penalties : [],
    categories: normalizeCategoryMap(raw.categories || {}),
  };
}

function normalizeWebParts(parts) {
  return parts.map((part) => ({
    name: String(part.name || ''),
    category: String(part.category || ''),
    damageMod: Number(part.damage_mod || 0),
    fireRateMod: Number(part.fire_rate_mod || 0),
  }));
}

function normalizeLegacyData(raw) {
  const categories = {};
  for (const group of Object.values(raw.Categories || {})) {
    for (const [name, idx] of Object.entries(group || {})) categories[name] = Number(idx);
  }

  return {
    cores: (raw.Data?.Cores || []).map((core) => {
      const damage = Array.isArray(core.Damage) ? core.Damage : [core.Damage, core.Damage];
      return {
        name: core.Name,
        category: core.Category,
        damage: Number(damage?.[0] || 0),
        damageEnd: Number(damage?.[1] || 0),
        fireRate: Number(core.Fire_Rate || 0),
      };
    }),
    magazines: (raw.Data?.Magazines || []).map((magazine) => ({
      name: magazine.Name,
      category: magazine.Category,
      magazineSize: Number(magazine.Magazine_Size || 0),
      damageMod: Number(magazine.Damage || 0),
      fireRateMod: Number(magazine.Fire_Rate || 0),
    })),
    barrels: mapLegacyPart(raw.Data?.Barrels || []),
    stocks: mapLegacyPart(raw.Data?.Stocks || []),
    grips: mapLegacyPart(raw.Data?.Grips || []),
    penalties: raw.Penalties || [],
    categories,
  };
}

function mapLegacyPart(parts) {
  return parts.map((part) => ({
    name: part.Name,
    category: part.Category,
    damageMod: Number(part.Damage || 0),
    fireRateMod: Number(part.Fire_Rate || 0),
  }));
}

function normalizeCategoryMap(categories) {
  return Object.fromEntries(Object.entries(categories).map(([name, index]) => [name, Number(index)]));
}

export function calculateTop(data, config) {
  return calculate(data, config).results;
}

export function calculate(data, config) {
  const normalizedConfig = normalizeConfig(config);
  const ranking = rankingFor(normalizedConfig);
  const results = [];
  const stats = {
    coresConsidered: 0,
    coresSkippedByCategory: 0,
    combinationsEvaluated: 0,
    combinationsFiltered: 0,
    resultsKept: 0,
  };

  for (const core of data.cores) {
    stats.coresConsidered += 1;

    if (!includeCategory(normalizedConfig.includeCategories, core.category)) {
      stats.coresSkippedByCategory += 1;
      continue;
    }

    const coreIdx = data.categories[core.category];
    if (coreIdx === undefined) continue;

    const magazines = topMagazines(data.magazines, core, coreIdx, normalizedConfig.partPoolPerType, data);
    const barrels = topParts(data.barrels, core, coreIdx, normalizedConfig.partPoolPerType, data);
    const stocks = topParts(data.stocks, core, coreIdx, normalizedConfig.partPoolPerType, data);
    const grips = topParts(data.grips, core, coreIdx, normalizedConfig.partPoolPerType, data);

    for (const magazine of magazines) {
      for (const barrel of barrels) {
        const magazineBarrelDamage = magazine.damageFactor * barrel.damageFactor;
        const magazineBarrelFireRate = magazine.fireRateFactor * barrel.fireRateFactor;

        for (const stock of stocks) {
          const baseDamageFactor = magazineBarrelDamage * stock.damageFactor;
          const baseFireRateFactor = magazineBarrelFireRate * stock.fireRateFactor;

          for (const grip of grips) {
            stats.combinationsEvaluated += 1;

            const metrics = evaluateMetrics(
              normalizedConfig,
              core,
              baseDamageFactor * grip.damageFactor,
              baseFireRateFactor * grip.fireRateFactor,
              magazine.item.magazineSize,
            );

            if (!metrics) continue;
            if (!passesFilters(normalizedConfig, metrics)) {
              stats.combinationsFiltered += 1;
              continue;
            }

            if (!canEnterTop(results, metrics, normalizedConfig.topN, ranking)) continue;

            insertTopResult(results, {
              core: core.name,
              magazine: magazine.item.name,
              barrel: barrel.item.name,
              stock: stock.item.name,
              grip: grip.item.name,
              ...metrics,
            }, normalizedConfig.topN, ranking);
          }
        }
      }
    }
  }

  stats.resultsKept = results.length;
  return { results, stats };
}

function normalizeConfig(config = {}) {
  return {
    topN: clampInteger(config.topN, 1, 100, 10),
    playerMaxHealth: positiveNumber(config.playerMaxHealth, 100),
    sortKey: Object.values(SORT_KEY).includes(config.sortKey) ? config.sortKey : SORT_KEY.TTK,
    priority: Object.values(PRIORITY).includes(config.priority) ? config.priority : PRIORITY.AUTO,
    includeCategories: Array.isArray(config.includeCategories) ? config.includeCategories.filter(Boolean) : [],
    partPoolPerType: clampInteger(config.partPoolPerType, 1, 50, 20),
    damageRange: normalizeRange(config.damageRange),
    damageEndRange: normalizeRange(config.damageEndRange),
    ttkSecondsRange: normalizeRange(config.ttkSecondsRange),
    dpsRange: normalizeRange(config.dpsRange),
  };
}

function normalizeRange(range) {
  const min = optionalFiniteNumber(range?.min);
  const max = optionalFiniteNumber(range?.max);
  return { min, max };
}

function optionalFiniteNumber(value) {
  if (value === '' || value === null || value === undefined) return null;
  const number = Number(value);
  return Number.isFinite(number) ? number : null;
}

function positiveNumber(value, fallback) {
  const number = Number(value);
  return Number.isFinite(number) && number > 0 ? number : fallback;
}

function clampInteger(value, min, max, fallback) {
  const number = Number.parseInt(value, 10);
  if (!Number.isFinite(number)) return fallback;
  return Math.min(max, Math.max(min, number));
}

function topParts(parts, core, coreIdx, limit, data) {
  return parts
    .map((part) => evaluatePart(part, core, coreIdx, data))
    .sort((a, b) => b.score - a.score)
    .slice(0, limit);
}

function topMagazines(magazines, core, coreIdx, limit, data) {
  return magazines
    .map((magazine) => {
      const evaluated = evaluatePart(magazine, core, coreIdx, data);
      return { ...evaluated, score: evaluated.score + magazine.magazineSize * 0.05 };
    })
    .sort((a, b) => b.score - a.score)
    .slice(0, limit);
}

function evaluatePart(part, core, coreIdx, data) {
  const penalty = core.name === part.name ? 0 : penaltyFor(data, coreIdx, part.category);
  const adjustedDamage = Number(part.damageMod || 0) * penalty;
  const adjustedFireRate = Number(part.fireRateMod || 0) * penalty;

  return {
    item: part,
    damageFactor: 1 + adjustedDamage / 100,
    fireRateFactor: 1 + adjustedFireRate / 100,
    score: adjustedDamage + adjustedFireRate * 0.6,
  };
}

function penaltyFor(data, coreIdx, category) {
  const partIdx = data.categories[category];
  if (partIdx === undefined) return 1;
  return Number(data.penalties?.[coreIdx]?.[partIdx] ?? 1);
}

function evaluateMetrics(config, core, damageFactor, fireRateFactor, magazineSize) {
  const damage = core.damage * damageFactor;
  const fireRate = core.fireRate * fireRateFactor;
  if (damage <= 0 || fireRate <= 0) return null;

  const shots = Math.ceil(config.playerMaxHealth / damage);
  const ttkSeconds = ((shots - 1) / fireRate) * 60;

  return {
    damage,
    damageEnd: core.damageEnd * damageFactor,
    fireRate,
    magazineSize,
    ttkSeconds,
    dps: (damage * fireRate) / 60,
  };
}

function passesFilters(config, metrics) {
  return inRange(config.damageRange, metrics.damage)
    && inRange(config.damageEndRange, metrics.damageEnd)
    && inRange(config.ttkSecondsRange, metrics.ttkSeconds)
    && inRange(config.dpsRange, metrics.dps);
}

function inRange(range, value) {
  if (range.min !== null && value < range.min) return false;
  if (range.max !== null && value > range.max) return false;
  return true;
}

function includeCategory(allowed, category) {
  return !allowed.length || allowed.some((candidate) => candidate.toLowerCase() === category.toLowerCase());
}

function rankingFor(config) {
  const priority = config.priority === PRIORITY.AUTO
    ? (config.sortKey === SORT_KEY.TTK ? PRIORITY.LOWEST : PRIORITY.HIGHEST)
    : config.priority;

  return { key: config.sortKey, priority };
}

function canEnterTop(results, metrics, topN, ranking) {
  if (results.length < topN) return true;
  return isBetter(metric(metrics, ranking.key), metric(results[results.length - 1], ranking.key), ranking.priority);
}

function insertTopResult(results, candidate, topN, ranking) {
  if (topN <= 0) return;

  const candidateMetric = metric(candidate, ranking.key);
  let insertAt = results.findIndex((result) => isBetter(candidateMetric, metric(result, ranking.key), ranking.priority));
  if (insertAt === -1) insertAt = results.length;
  results.splice(insertAt, 0, candidate);
  if (results.length > topN) results.pop();
}

function isBetter(left, right, priority) {
  return priority === PRIORITY.HIGHEST ? left > right : left < right;
}

function metric(result, key) {
  switch (key) {
    case SORT_KEY.DPS: return result.dps;
    case SORT_KEY.DAMAGE: return result.damage;
    case SORT_KEY.DAMAGE_END: return result.damageEnd;
    case SORT_KEY.FIRE_RATE: return result.fireRate;
    case SORT_KEY.MAGAZINE: return result.magazineSize;
    case SORT_KEY.TTK:
    default: return result.ttkSeconds;
  }
}

export function formatResults(results) {
  if (!results.length) return 'No results found.';
  return results.map((result, index) => [
    `#${index + 1}`,
    `Core: ${result.core}`,
    `Magazine: ${result.magazine}`,
    `Barrel: ${result.barrel}`,
    `Stock: ${result.stock}`,
    `Grip: ${result.grip}`,
    `Damage: ${result.damage.toFixed(3)}`,
    `Damage End: ${result.damageEnd.toFixed(3)}`,
    `Fire Rate: ${result.fireRate.toFixed(3)}`,
    `Magazine: ${result.magazineSize.toFixed(0)}`,
    `TTK: ${result.ttkSeconds.toFixed(3)}s`,
    `DPS: ${result.dps.toFixed(3)}`,
  ].join('\n')).join('\n\n');
}

export { SORT_KEY, PRIORITY };

File: src/engine.rs
use std::{cmp::Ordering, collections::HashMap};

#[derive(Debug, Clone, Copy, Default, PartialEq)]
pub struct NumericRange {
    pub min: Option<f64>,
    pub max: Option<f64>,
}

impl NumericRange {
    pub const fn new(min: Option<f64>, max: Option<f64>) -> Self {
        Self { min, max }
    }

    pub fn contains(self, value: f64) -> bool {
        if self.min.is_some_and(|min| value < min) {
            return false;
        }
        if self.max.is_some_and(|max| value > max) {
            return false;
        }
        true
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SortKey {
    Ttk,
    Dps,
    Damage,
    DamageEnd,
    FireRate,
    Magazine,
}

impl SortKey {
    pub fn parse(value: &str) -> anyhow::Result<Self> {
        match value.to_ascii_lowercase().as_str() {
            "ttk" => Ok(Self::Ttk),
            "dps" => Ok(Self::Dps),
            "damage" => Ok(Self::Damage),
            "damageend" => Ok(Self::DamageEnd),
            "firerate" => Ok(Self::FireRate),
            "magazine" => Ok(Self::Magazine),
            _ => anyhow::bail!("Invalid sort key: {value}"),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SortPriority {
    Highest,
    Lowest,
    Auto,
}

impl SortPriority {
    pub fn parse(value: &str) -> anyhow::Result<Self> {
        match value.to_ascii_lowercase().as_str() {
            "highest" => Ok(Self::Highest),
            "lowest" => Ok(Self::Lowest),
            "auto" => Ok(Self::Auto),
            _ => anyhow::bail!("Invalid priority: {value}"),
        }
    }
}

#[derive(Debug, Clone)]
pub struct Config {
    pub data_path: String,
    pub top_n: usize,
    pub player_max_health: f64,
    pub sort_key: SortKey,
    pub priority: SortPriority,
    pub include_categories: Vec<String>,
    pub damage_range: NumericRange,
    pub damage_end_range: NumericRange,
    pub ttk_seconds_range: NumericRange,
    pub dps_range: NumericRange,
    pub part_pool_per_type: usize,
}

impl Default for Config {
    fn default() -> Self {
        Self {
            data_path: "Data/FullData.sqlite3".into(),
            top_n: 10,
            player_max_health: 100.0,
            sort_key: SortKey::Ttk,
            priority: SortPriority::Auto,
            include_categories: Vec::new(),
            damage_range: NumericRange::default(),
            damage_end_range: NumericRange::default(),
            ttk_seconds_range: NumericRange::default(),
            dps_range: NumericRange::default(),
            part_pool_per_type: 20,
        }
    }
}

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct CalculationStats {
    pub cores_considered: usize,
    pub cores_skipped_by_category: usize,
    pub combinations_evaluated: u64,
    pub combinations_filtered: u64,
    pub results_kept: usize,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Core {
    pub name: String,
    pub category: String,
    pub damage: f64,
    pub damage_end: f64,
    pub fire_rate: f64,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Magazine {
    pub name: String,
    pub category: String,
    pub magazine_size: f64,
    pub reload_time: f64,
    pub damage_mod: f64,
    pub fire_rate_mod: f64,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Part {
    pub name: String,
    pub category: String,
    pub damage_mod: f64,
    pub fire_rate_mod: f64,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ResultRow {
    pub core: String,
    pub magazine: String,
    pub barrel: String,
    pub stock: String,
    pub grip: String,
    pub damage: f64,
    pub damage_end: f64,
    pub fire_rate: f64,
    pub magazine_size: f64,
    pub ttk_seconds: f64,
    pub dps: f64,
}

#[derive(Debug, Clone, PartialEq)]
pub struct DataSet {
    pub cores: Vec<Core>,
    pub magazines: Vec<Magazine>,
    pub barrels: Vec<Part>,
    pub grips: Vec<Part>,
    pub stocks: Vec<Part>,
    pub penalties: Vec<Vec<f64>>,
    pub categories: HashMap<String, usize>,
}

pub struct Engine<'a> {
    data: &'a DataSet,
}

impl<'a> Engine<'a> {
    pub const fn new(data: &'a DataSet) -> Self {
        Self { data }
    }

    pub fn calculate_top(&self, config: &Config) -> (Vec<ResultRow>, CalculationStats) {
        let mut results = Vec::new();
        let mut stats = CalculationStats::default();

        for core in &self.data.cores {
            stats.cores_considered += 1;
            if !include_category(&config.include_categories, &core.category) {
                stats.cores_skipped_by_category += 1;
                continue;
            }

            let Some(&core_idx) = self.data.categories.get(&core.category) else {
                continue;
            };

            let magazines = self.top_magazines(core, core_idx, config.part_pool_per_type);
            let barrels = self.top_parts(&self.data.barrels, core, core_idx, config.part_pool_per_type);
            let stocks = self.top_parts(&self.data.stocks, core, core_idx, config.part_pool_per_type);
            let grips = self.top_parts(&self.data.grips, core, core_idx, config.part_pool_per_type);

            for mag in magazines {
                for barrel in &barrels {
                    for stock in &stocks {
                        for grip in &grips {
                            stats.combinations_evaluated += 1;
                            let Some(candidate) = self.build_result(config, core, core_idx, mag, barrel, stock, grip) else {
                                continue;
                            };
                            if !passes_filters(config, &candidate) {
                                stats.combinations_filtered += 1;
                                continue;
                            }
                            self.push_top(&mut results, candidate, config);
                        }
                    }
                }
            }
        }

        stats.results_kept = results.len();
        (results, stats)
    }

    fn push_top(&self, results: &mut Vec<ResultRow>, candidate: ResultRow, config: &Config) {
        if config.top_n == 0 {
            return;
        }

        if results.len() < config.top_n {
            results.push(candidate);
        } else if better(&candidate, results.last().expect("non-empty top list"), config) {
            *results.last_mut().expect("non-empty top list") = candidate;
        } else {
            return;
        }

        results.sort_by(|a, b| {
            if better(a, b, config) {
                Ordering::Less
            } else if better(b, a, config) {
                Ordering::Greater
            } else {
                Ordering::Equal
            }
        });
    }

    fn build_result(
        &self,
        config: &Config,
        core: &Core,
        core_idx: usize,
        mag: &Magazine,
        barrel: &Part,
        stock: &Part,
        grip: &Part,
    ) -> Option<ResultRow> {
        let damage_parts = [
            (&mag.name, &mag.category, mag.damage_mod),
            (&barrel.name, &barrel.category, barrel.damage_mod),
            (&stock.name, &stock.category, stock.damage_mod),
            (&grip.name, &grip.category, grip.damage_mod),
        ];
        let fire_rate_parts = [
            (&mag.name, &mag.category, mag.fire_rate_mod),
            (&barrel.name, &barrel.category, barrel.fire_rate_mod),
            (&stock.name, &stock.category, stock.fire_rate_mod),
            (&grip.name, &grip.category, grip.fire_rate_mod),
        ];

        let damage_mult = self.percent_multiplier(core, core_idx, &damage_parts);
        let fire_rate_mult = self.percent_multiplier(core, core_idx, &fire_rate_parts);
        let damage = core.damage * damage_mult;
        let fire_rate = core.fire_rate * fire_rate_mult;
        if damage <= 0.0 || fire_rate <= 0.0 {
            return None;
        }

        let shots = (config.player_max_health / damage).ceil();
        let ttk_seconds = ((shots - 1.0) / fire_rate) * 60.0;

        Some(ResultRow {
            core: core.name.clone(),
            magazine: mag.name.clone(),
            barrel: barrel.name.clone(),
            stock: stock.name.clone(),
            grip: grip.name.clone(),
            damage,
            damage_end: core.damage_end * damage_mult,
            fire_rate,
            magazine_size: mag.magazine_size,
            ttk_seconds,
            dps: damage * fire_rate / 60.0,
        })
    }

    fn percent_multiplier(&self, core: &Core, core_idx: usize, parts: &[(&String, &String, f64)]) -> f64 {
        parts.iter().fold(1.0, |mult, (name, category, raw_mod)| {
            let penalty = self.penalty_for(core_idx, category);
            mult * (1.0 + adjusted_mod(*raw_mod, &core.name, name, penalty) / 100.0)
        })
    }

    fn top_parts<'b>(&self, pool: &'b [Part], core: &Core, core_idx: usize, max_count: usize) -> Vec<&'b Part> {
        let mut parts: Vec<_> = pool.iter().collect();
        parts.sort_by(|a, b| {
            self.part_score(core, core_idx, &b.name, &b.category, b.damage_mod, b.fire_rate_mod)
                .partial_cmp(&self.part_score(core, core_idx, &a.name, &a.category, a.damage_mod, a.fire_rate_mod))
                .unwrap_or(Ordering::Equal)
        });
        parts.truncate(max_count);
        parts
    }

    fn top_magazines<'b>(&'b self, core: &Core, core_idx: usize, max_count: usize) -> Vec<&'b Magazine> {
        let mut magazines: Vec<_> = self.data.magazines.iter().collect();
        magazines.sort_by(|a, b| {
            let score_a = self.part_score(core, core_idx, &a.name, &a.category, a.damage_mod, a.fire_rate_mod)
                + a.magazine_size * 0.05;
            let score_b = self.part_score(core, core_idx, &b.name, &b.category, b.damage_mod, b.fire_rate_mod)
                + b.magazine_size * 0.05;
            score_b.partial_cmp(&score_a).unwrap_or(Ordering::Equal)
        });
        magazines.truncate(max_count);
        magazines
    }

    fn part_score(&self, core: &Core, core_idx: usize, part_name: &str, category: &str, damage_mod: f64, fire_rate_mod: f64) -> f64 {
        let penalty = self.penalty_for(core_idx, category);
        adjusted_mod(damage_mod, &core.name, part_name, penalty)
            + adjusted_mod(fire_rate_mod, &core.name, part_name, penalty) * 0.6
    }

    fn penalty_for(&self, core_idx: usize, category: &str) -> f64 {
        let Some(&part_idx) = self.data.categories.get(category) else {
            return 1.0;
        };
        self.data.penalties
            .get(core_idx)
            .and_then(|row| row.get(part_idx))
            .copied()
            .unwrap_or(1.0)
    }
}

fn adjusted_mod(raw: f64, core_name: &str, part_name: &str, penalty: f64) -> f64 {
    if core_name == part_name { 0.0 } else { raw * penalty }
}

fn passes_filters(config: &Config, result: &ResultRow) -> bool {
    config.damage_range.contains(result.damage)
        && config.damage_end_range.contains(result.damage_end)
        && config.ttk_seconds_range.contains(result.ttk_seconds)
        && config.dps_range.contains(result.dps)
}

fn include_category(allowed: &[String], category: &str) -> bool {
    allowed.is_empty() || allowed.iter().any(|candidate| candidate.eq_ignore_ascii_case(category))
}

fn better(a: &ResultRow, b: &ResultRow, config: &Config) -> bool {
    let left = metric(a, config.sort_key);
    let right = metric(b, config.sort_key);
    let priority = match config.priority {
        SortPriority::Auto if config.sort_key == SortKey::Ttk => SortPriority::Lowest,
        SortPriority::Auto => SortPriority::Highest,
        explicit => explicit,
    };
    match priority {
        SortPriority::Highest => left > right,
        SortPriority::Lowest => left < right,
        SortPriority::Auto => unreachable!(),
    }
}

fn metric(result: &ResultRow, key: SortKey) -> f64 {
    match key {
        SortKey::Ttk => result.ttk_seconds,
        SortKey::Dps => result.dps,
        SortKey::Damage => result.damage,
        SortKey::DamageEnd => result.damage_end,
        SortKey::FireRate => result.fire_rate,
        SortKey::Magazine => result.magazine_size,
    }
}

pub fn write_results(results: &[ResultRow]) -> String {
    let mut out = String::new();
    for (idx, r) in results.iter().enumerate() {
        use std::fmt::Write;
        let _ = writeln!(out, "#{}", idx + 1);
        let _ = writeln!(out, " Core: {}", r.core);
        let _ = writeln!(out, " Magazine: {}", r.magazine);
        let _ = writeln!(out, " Barrel: {}", r.barrel);
        let _ = writeln!(out, " Stock: {}", r.stock);
        let _ = writeln!(out, " Grip: {}", r.grip);
        let _ = writeln!(out, " Damage: {:.3}", r.damage);
        let _ = writeln!(out, " Damage End: {:.3}", r.damage_end);
        let _ = writeln!(out, " Fire Rate: {:.3}", r.fire_rate);
        let _ = writeln!(out, " TTK: {:.3}s", r.ttk_seconds);
        let _ = writeln!(out, " DPS: {:.3}\n", r.dps);
    }
    out
}
